doc: ADR for Limit LMS Modulestore access (#27620)
edx/edx-platform: docs/decisions/0010-limit-modulestore-use-in-lms.rst

> Limit LMS Modulestore access to the courseware app

Reviewer: The courseware app is called out as the big exception here, but it isn't mentioned anywhere why.

I think I know why (rendering the unit, yeah?) but could you give it a sentence or two in the ADR? And what do you see as the big blocker towards resolving it?

Author: Ah right, makes sense. Will add.

--------------------------------------------------

Context
=======

Django apps deployed in the LMS sometimes want to make queries about course team
authored content, like sequences, units, and problems. To do this, they often
use the Modulestore, which can return parts of the course as a graph of XBlocks.
Unfortunately, the Modulestore is a large, complex system that is shared by many
features. This has been the cause of many bugs and performance issues over the
years. Modulestore performance has been gradually improved over the years, but
at the cost of adding even more complexity.

Decisions
=========

1. New features will be implemented without making calls to the Modulestore in
   the LMS.
2. Teams should opportunistically remove LMS Modulestore dependencies in

>    existing features as they are being worked on.

Reviewer: Does this include indirect dependencies? As I understand it, calling `has_access` with a usage key dives into the modulestore.

Do we want to warn against using that or any other utility functions that are big modulestore offenders? Or are you hoping just to focus on direct dependencies here, and tackle utilities later?

Author: Yeah, I'm hoping to hit the main apps and then deal with the extended permissions stuff elsewhere as a separate effort. I think it's a much more straightforward ask to ask people to remove modulestore usage from the things their team owns, while asking them to remove access checks from the bowels of `has_access` is a much harder sell. We will have to cross that bridge (and burn it to cinders) at some point though.

> 3. Apps may still access the Modulestore in Studio.

Author: Good point. I think the answer is "both", but I need to think about it a bit...

  predictable.

**Tests are easier to write and run faster.**
  CourseOverviews and UserCourseOutlineData objects are much easier to create
  and mock for testing purposes than a tree of XBlocks, and do not suffer from
  complicated publishing rules. Using the modulestore also imposes a substantial
  performance penalty when creating and modifying courses, making modulestore
  access the dominant part of the edx-platform test suite running time.

**Applications are more resilient to user-facing failures.**
  Many features today are at least partially implemented in Modulestore and the
  XBlocks that it returns. Changes in those features can cause bugs that ripple
  out into completely unrelated features being developed by other teams.

  If an LMS application queries the Modulestore when responding to a user
  request, then unexpected failures in shared Modulestore code will cause
  user-facing errors. A feature that reads Modulestore data at the time of
  course publish and pushes it into its own data model will still be affected by
  Modulestore-related bugs, but it will result in stale data and not an entirely
  broken experience.

> **We can reduce the size of the edx-platform monolith.**

Reviewer: What do you think about mentioning LMS/CMS decoupling, either as a note within this goal or its own goal?

Apps that need course outline data and metadata about sequences can make use
of the Learning Sequences public API from the
``openedx.core.djangoapps.content.learning_sequences.api`` package. A couple of
important caveats to the functions advertised here:

* Old Mongo courses are not supported. These are soon-to-be-removed courses with
  course keys of the format "Org/Course/Run". All course keys that begin with
  "course-v1:" or "ccx-v1:" are supported.
* This is a new API, and new API functions and data will be added over time.

Advanced Use Cases
******************

Sometimes the data an application needs just isn't available in a performant way
in the LMS. When this happens, you'll need to create your own model for this
data in your application, and then push new data to it during the course publish
process. To demonstrate how this can work, let's walk through the implementation
of Learning Sequences, which follows this general architectural pattern.

Studio Process
^^^^^^^^^^^^^^

The code to extract data from the Modulestore and convert it into a format that
the Learning Sequences app understands lives in the ``cms.djangoapps.contentstore``
package. This has several pieces:

> Data Extraction

Reviewer: The rendered RST doesn't make it clear that these are sub-headings under Studio process. Will it let you do another level of heading nesting?

  (it's a BSON object).

Celery Task
  ``cms.djangoapps.contentstore.outlines.tasks.update_outline_from_modulestore_task``

  This is a simple celery ``@shared_task`` that wraps the call to
  ``update_outline_from_modulestore``. It's critical to use celery to do this
  work asynchronously. Even if your code seems to work quickly enough to run
  in-process, courses can often use obscure features that can drastically
  increase the time it takes to get data out, and you will almost certainly not
  be able to comprehensively test for all those situations.

  *You must be aggressive about alerting on task failures*. Publishes are
  infrequent enough to make it so that certain content-dependent errors will not
  trigger most error rate alerts. You have to be extremely sensitive to outright
  failures in your task because you may be blocking the publish for a course.

Signal Handler
  ``cms.djangoapps.contentstore.outlines.signals.handlers.listen_for_course_publish``

  This is a centralized location where Studio does its post-publish data pushes,
  but you can also make a separate handler that listens for the same

>   ``course_published`` signal. It's main task is to do some logging and queue

Reviewer 2: `It's` -> `Its`

  the celery task.

Management Task

>   ``cms.djangoapps.contentstore.management.commands.backfill_course_outlines.py``
>   ``cms.djangoapps.contentstore.management.commands.update_course_outline.py``

Reviewer: ```suggestion
  ``cms.djangoapps.contentstore.management.commands.backfill_course_outlines``
  ``cms.djangoapps.contentstore.management.commands.update_course_outline``
```

  Management commands to backfill a group of course outlines or to update one
  particular command. A few things to note:

  1. These commands live in the Studio process, because they are invoking code
     that will query the Modulestore.
  2. The backfill command launches a new celery task for every individual
     course. This is both to control memory usage (successive Modulestore access
     across courses will leak a lot of memory), as well as to make it easier to
     see which courses are taking longer and/or causing errors.
  3. In the long term, you will want a way to trigger backfills from the Django

>      admin, so that you don't need to file a support ticket every time.

Reviewer: Once the app-permissions work is done and there's a admin action for backfilling learning sequences, let's come back and update this conversion guide. Ideally, future devs in this situation won't even need to make a management command.

LMS Process
^^^^^^^^^^^

> In the LMS process, your feature should be completely free of Modulestore
> dependencies. In fact, your app should be *entirely* free of Modulestore
> dependencies. All of that code is in the Studio section described above. By the
> time your LMS request is happening, your app is only looking at its own data
> models, or one of the performant Modulestore-alternative APIs.

Reviewer: This paragraph is a little confusing in its current wording.

```suggestion
In the LMS process, your feature should not use the Modulestore at all.
Your LMS app's code should be entirely free of Modulestore dependencies.
All of the Modulestore-facing code described above should live in the ``./cms/`` source tree and run in the Studio process.
By the time your LMS request is happening, your app is only looking at its own data
models, or one of the performant Modulestore-alternative APIs.
```

You should not allow the LMS process to overwrite models written to by the
course publishing process, and you should absolutely not let the LMS push data
back into the Modulestore. If your application needs to be able to override data
that comes from publishing, have two separate models–one that's only ever
updated by course content publishing, and one that's read/write from the LMS.
When answering queries, your app can look at both models. The edx-when app works
in this way, capturing start and due date information from the Modulestore, but
then applying student-specific overrides when serving requests in the LMS. For

> more background on this topic, please see ADR 0005-studio-lms-subdomain-boundaries.rst.

Reviewer: ```suggestion
more background on this topic, please see `ADR 5: LMS Subdomain Boundaries <./docs/decisions/0005-studio-lms-subdomain-boundaries.rst>`_ .
```
